### PR TITLE
Improve setting signalsWithMasterHandlers and fix omrsig_protect

### PR DESCRIPTION
**Pull request by babsingh**
Currently, signalsWithMasterHandlers is set at two places:
omrsig_register_os_handler and registerMasterHandlers.

With this changeset, signalsWithMasterHandlers will only be set within
registerSignalHandlerWithOS after an OS handler has been successfully
registered.

A write barrier is issued before updating signalsWithHandlers and
signalsWithMasterHandlers within registerSignalHandlerWithOS. The write
barrier will allow us to read signalsWithHandlers and
signalsWithMasterHandlers without acquiring registerHandlerMonitor.

omrsig_protect serves asynchronous signals, and OS handlers for
asynchronous signals are generally registered once at startup. Now
onwards, signalsWithMasterHandlers is read without acquiring
registerHandlerMonitor within omrsig_protect. This serves two purposes. 

First, omrsig_protect won't acquire registerHandlerMonitor if a handler
is already registered. From past performance numbers, this would save
1828 instructions from the newDirectByteBuffer pathlength. 

Second, thread->monitor value ends up being NULL if the thread gets
aborted during the call to
omrthread_monitor_enter(registerHandlerMonitor). This impacts the
behavior of omrthread.c::interrupt_waiting_server. In OpenJ9,
intermittent segmentation faults are observed during calls to
Java_java_lang_Thread_interruptImpl. By not acquiring
registerHandlerMonitor if an OS handler is already registered,
omrsig_protect avoids conflict with the thread library and prevents
segmentation fault in OpenJ9.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
fyi - @pshipton this is the fix for the gpf issue. i am testing changes in personal builds. i will remove the WIP after testing.

File: port/unix/omrsignal.c [hunk 1]
```diff
@@ -1274,10 +1258,22 @@ registerSignalHandlerWithOS(OMRPortLibrary *portLibrary, uint32_t portLibrarySig
 		}
 	}
 
+	issueWriteBarrier();
+
 	/* Set the portLibrarySignalNo bit in signalsWithHandlers to record successful registration
 	 * of the handler. */
 	signalsWithHandlers |= portLibrarySignalNo;
```
Comment (pshipton) [inline port/unix/omrsignal.c]: Seems these updates need to be atomic. Otherwise without a monitor bits could be lost.

Comment (babsingh) [reply port/unix/omrsignal.c]: a monitor is always acquired before calling registerSignalHandlerWithOS.

**Comment by babsingh** [comment]
Removed the WIP tag. Personal builds and local testing have succeeded.

**Comment by charliegracie** [comment]
@genie-omr build all

File: port/unix/omrsignal.c (hunk 1 above)
Comment (charliegracie) [reply port/unix/omrsignal.c]: Is there a lock being held at this point while the flags are being modified?

Comment (babsingh) [reply port/unix/omrsignal.c]: calls to `registerSignalHandlerWithOS` always acquire `registerHandlerMonitor`.

**Pull request merged**